### DISCUSSION TITLE
Library error: connection closed unexpectedly - Potential login failure.

### DISCUSSION THREAD

**Discussion by IYABAO** [pdezwart/php-amqp#445] [Q&A]
Occasional problems occur when I produce messages in production environment. 
And it come up very often.

Library error: connection closed unexpectedly - Potential login failure.

I googled, and checked all the related issues in Github and StackOverflow, but don't find the related solution.
As I said the problem is occasional. My login info is ok. When I catch the exception and retry it. It works well.

`bash-5.1# php --ri amqp

amqp

Version => 1.11.0
Revision => release
Compiled => Nov 29 2022 @ 06:25:39
AMQP protocol version => 0-9-1
librabbitmq version => 0.11.0
Default max channels per connection => 256
Default max frame size => 131072
Default heartbeats interval => 0

Directive => Local Value => Master Value
amqp.host => localhost => localhost
amqp.vhost => / => /
amqp.port => 5672 => 5672
amqp.timeout => no value => no value
amqp.read_timeout => 0 => 0
amqp.write_timeout => 0 => 0
amqp.connect_timeout => 0 => 0
amqp.rpc_timeout => 0 => 0
amqp.login => guest => guest
amqp.password => guest => guest
amqp.auto_ack => 0 => 0
amqp.prefetch_count => 3 => 3
amqp.prefetch_size => 0 => 0
amqp.global_prefetch_count => 0 => 0
amqp.global_prefetch_size => 0 => 0
amqp.channel_max => 256 => 256
amqp.frame_max => 131072 => 131072
amqp.heartbeat => 0 => 0
amqp.cacert => no value => no value
amqp.cert => no value => no value
amqp.key => no value => no value
amqp.verify => 1 => 1
amqp.sasl_method => 0 => 0`

I checked the RabbitMQ node statistics. "File descriptors", "Socket descriptors", "Erlang processes". The node is health enough.

**Comment by lyrixx** [comment]
Can it be a network issue?

**Comment by lstrojny** [comment]
The error message is admittedly not very specific but rather a general "something went wrong on a TCP level". I've seen this error message sometimes with a load balancer that closes inactive  connection after some time. You could try to set a heartbeat interval to see if it goes away. Additionally check your network stack for timeouts (I know, this is not very specific advice).

**Comment by IYABAO** [comment]
> Can it be a network issue?

A high probability. Just can't locate the problem and hard to start debugging. Submit this issue not to report a bug but to find some buys who face the same situation can share some experience. Maybe some config setting or params when build the connection. Thanks for your time.

**Comment by IYABAO** [comment]
> The error message is admittedly not very specific but rather a general "something went wrong on a TCP level". I've seen this error message sometimes with a load balancer that closes inactive connection after some time. You could try to set a heartbeat interval to see if it goes away. Additionally check your network stack for timeouts (I know, this is not very specific advice).

Thanks for your time and advice.  I tried to adjust the heartbeat and timeout setting in dev environment. When it comes nearly 1000+ connection, it shows other errors.
`
Broken pipe or closed connection
stream_socket_client(): unable to connect to tcp://myip:5672 (Operation timed out)
Missed server heartbeat
`
Can I get some logs in RabbitMQ server when `connection closed unexpectedly - Potential login failure.` occur?  Or I just have to analysis the networks and connection condition in the client side?

**Comment by lstrojny** [comment]
Moved to discussions
